backends/dpdk/dpdkArch.cpp

```diff
@@ -3035,4 +3035,153 @@ std::vector<std::pair<cstring, const IR::Type *>>
+                ipsecHdrAdded = true;
+                new_objs->push_back(ipsecHeader);
+            }
+        } else if (obj->is<IR::P4Parser>()) {
+            if (!ipsecPortRegAdded) {
+                ipsecPortRegAdded = true;
+                // Add all 4 ipsec port register declarations before parser
+                auto decls = addRegDeclInstance(registerInstanceNames);
+                for (auto d : *decls) new_objs->push_back(d);
+            }
+        }
+        new_objs->push_back(obj);
+    }
+    if (!(ipsecHdrAdded && ipsecPortRegAdded))
+        BUG("Missing platform header/IPSec port Registers needed for IPSec accelerator");
+    program->objects = *new_objs;
+    return program;
+}
+
+// Create an instance of ipsec_hdr in the header struct
+const IR::Node *InsertReqDeclForIPSec::preorder(IR::Type_Struct *s) {
+    if (!is_ipsec_used) return s;
+    if (s->name.name == structure->header_type) {
+        s->fields.push_back(
+            new IR::StructField(IR::ID("ipsec_hdr"), new IR::Type_Name(IR::ID("platform_hdr_t"))));
+    }
+    return s;
+}
+
+const IR::Node *InsertReqDeclForIPSec::preorder(IR::P4Control *c) {
+    if (!is_ipsec_used) return c;
+
+    auto insertIpsecHeader = [this](IR::P4Control *c) {
+        IR::IndexedVector<IR::StatOrDecl> stmt;
+        auto pkt_out = c->type->applyParams->parameters.at(0);
+        auto args = new IR::Vector<IR::Argument>();
+        structure->ipsec_header =
+            new IR::Member(new IR::PathExpression(IR::ID("h")), IR::ID("ipsec_hdr"));
+        args->push_back(new IR::Argument(structure->ipsec_header));
+        auto typeArgs = new IR::Vector<IR::Type>();
+        typeArgs->push_back(new IR::Type_Name(IR::ID(newHeaderName)));
+        auto regRdArgs = new IR::Vector<IR::Argument>();
+
+        // Register index is always 0 as the ipsec port registers are all single location registers
+        regRdArgs->push_back(
+            new IR::Argument(new IR::Constant(IR::Type::Bits::get(32), IPSEC_PORT_REG_INDEX)));
+        /*  Set the ipsec output port based on packet direction and emit ipsec_hdr
+         *  if (ipsec_hdr.isValid())
+                 port_out = (direction == NET_TO_HOST) ?
+                            ipsec_inbound_port_out: ipsec_outbound_port_out;
+                 emit(ipsec_hdr);
+        */
+        auto regRdOutInbound = new IR::MethodCallExpression(
+            new IR::Member(new IR::PathExpression(IR::ID("ipsec_port_out_inbound")),
+                           IR::ID("read")),
+            regRdArgs);
+        auto regRdOutOutbound = new IR::MethodCallExpression(
+            new IR::Member(new IR::PathExpression(IR::ID("ipsec_port_out_outbound")),
+                           IR::ID("read")),
+            regRdArgs);
+
+        auto portOutInBound = new IR::AssignmentStatement(
+            c->body->srcInfo,
+            new IR::Member(new IR::PathExpression(IR::ID("m")),
+                           IR::ID("pna_main_output_metadata_output_port")),
```

Comment:
one of these two assignments must be wrong, since they write to the same field

Reply:
These assignments are placed in true and false blocks of "if-else"
corresponds to 
   ```
  if (direction == NET_TO_HOST)
	     m.pna_main_output_metadata_output_port = ipsec_inbound_port_out
     else
	     m.pna_main_output_metadata_output_port = ipsec_outbound_port_out
```

Reply:
it would be clearer to factor out the common subexpression. But it makes sense.

Reply:
Updated comments and refactored common subexpression.